### PR TITLE
Capture Cplex specific errors and raise a unified `optlang.SolverError` [WIP]

### PR DESCRIPTION
**Pull request by Midnighter**
When working with different optlang solver interfaces error handling should be agnostic. So far this is implemented for Cplex and Gurobi should come next. As far as I'm aware GLPK raises no errors but maybe surprising behavior should still be handled by raising from optlang. I'm not sure.

So far this pull request also implements test cases for accessing dual values on MIP problems. These should go into a separate branch/PR, I suppose.

Tests cannot pass at the moment since I introduce two failing tests due to the above.

### CODE REVIEW

File: optlang/tests/test_cplex_interface.py [hunk 1]
```diff
@@ -497,6 +499,98 @@ def test_qp_non_convex(self):
             model.optimize()
             self.assertAlmostEqual(model.objective.value, 2441.999999971)
 
+    class InfeasibleMIPTestCase(unittest.TestCase):
```
Comment (KristianJensen) [inline optlang/tests/test_cplex_interface.py]: Couldn't this be combined with the other test case?

File: optlang/cplex_interface.py [hunk 2]
```diff
@@ -31,14 +31,17 @@
 
 log = logging.getLogger(__name__)
 
+from future.utils import raise_from
```
Comment (Midnighter) [inline optlang/cplex_interface.py]: `future` is built on top of six and I find it a lot nicer in many cases but sure, I can drop it.

File: optlang/tests/test_cplex_interface.py (hunk 1 above)
Comment (Midnighter) [inline optlang/tests/test_cplex_interface.py]: I could drop those cases that are tested by accessing dual values in MIP already, true. However, I want to test here that all Cplex errors are adequately caught and reraised as `optlang.SolverError`s. That behavior is different between MIPs and LPs. Maybe I should extend this to LPs and QPs, too.

Comment (KristianJensen) [inline optlang/tests/test_cplex_interface.py]: I agree that it's good to test this as well, and I just meant that these tests could simply be added to the other test case.

Comment (Midnighter) [inline optlang/tests/test_cplex_interface.py]: I'd rather keep them separate since the intention of the tests is something different.

**Comment by Midnighter** [comment]
Closing due to #96 which is a polished version of this branch.

**Pull request closed without merging**